### PR TITLE
Performance: Optimization for master-key authentication

### PR DESCRIPTION
**Pull request by kirankumarkolli**
Key optimizations cover yields 0.3% CPU
- Reusing the mac instances
- Use object == to avoid hash computation

Baseline: 
image

With optimization: 
image

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java [hunk 1]
```diff
@@ -237,38 +239,44 @@ private String generateKeyAuthorizationSignatureNew(RequestVerb verb, String res
...
-            //  Master key hasn't changed, return the cloned mac instance
-            try {
-                return (Mac)this.macInstance.clone();
-            } catch (CloneNotSupportedException e) {
-                throw new IllegalStateException(e);
+    private MacPool.ReUsableMac getReUseableMacInstance() {
+
+        // Java == operator is reference equals not content
+        // leveraging reference comparison avoid hash computation
+        if (this.currentCredentialKey != this.credential.getKey()) {
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java:260]: SpotBugs cribbing, @kushagraThapar, @moderakh how to suppress this warning?

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java:260]: @kirankumarkolli  - you need to update this file with the proper spotbug issue which is failing here : https://github.com/Azure/azure-sdk-for-java/blob/master/eng/code-quality-reports/src/main/resources/spotbugs/spotbugs-exclude.xml

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/MacPool.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation;
+
+import javax.crypto.Mac;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+class MacPool {
+    final Mac macInstance;
+    final ConcurrentLinkedQueue<Mac> pool;
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/MacPool.java:11]: Using MacPool is a very nice change.

however now mac.clone() overhead is replaced by a synchrnozation lock. most likely better perf. but do we know how much perf gain?

You can do micro-benchmarking using jmh (just for the authorization part) I have the jmh project template here:
https://github.com/moderakh/azure-cosmosdb-benchmark/tree/master/jmh

Comment (kirankumarkolli) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/MacPool.java:11]: Won't this benchmark needs package to be published?

Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/MacPool.java:11]: no, you don't need to publish the package. There are two options:

first option)
 on the java SDK repo install the package in your local maven repo using following command:
```mvn -e  -Dgpg.skip -Dmaven.javadoc.skip=true -Dspotbugs.skip=true  -Dcheckstyle.skip=true  -Drevapi.skip=true  -DskipTests -pl ,com.azure:azure-cosmos  -am clean install```
and in the jmh project just update the pom to use the latest version installed in your local maven repo.
example for this option in jmh project:
https://github.com/moderakh/azure-cosmosdb-benchmark/blob/master/jmh/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdTokenStreamJMH.java

second option)
if the area to be tested is small (just copy that code (e.g, MacPool) to jmh project itself and micro benchmark just the copied code.
example for this option:
copied code:
https://github.com/moderakh/azure-cosmosdb-benchmark/blob/master/jmh/src/main/java/moderakh/ConsistencyLevel.java
micro-benchmark test:
https://github.com/moderakh/azure-cosmosdb-benchmark/blob/master/jmh/src/main/java/moderakh/ExpandableStringEnumJMH.java

Comment (kirankumarkolli) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/MacPool.java:11]: Based on offline discussion resolving this comment.

**Pull request merged**